### DISCUSSION TITLE
Why do you build a DIY edrum?

### DISCUSSION THREAD

**Discussion by corrados** [corrados/edrumulus#4] [General]
I am interested to know why you spend time on developing/building a do-it-yourself edrum.

The trigger to create Edrumulus was that I was disappointed about the fact that entry level Roland edrum modules do not support positional sensing. If you want to have positional sensing, you have to spend a lot more money. Then I recorded a edrum pad signal with the soundcard of my PC and analyzed the signal in Octave. After some evaluation and testing I found a simple way to estimate the position. Edrumulus was born :-).

Actually, I do not have the need to have another edrum module. Right now I have a TD-6, a TD-17 and a TD-27. I even have a Roland digital snare pad. So, for me it is just the fun of exploring new algorithms and see what you can do with a very cheap hardware like the ESP32 or Teensy.

Let us know what is your motivation for building DIY edrums.

**Comment by SpintroniK** [comment]
Hi Volker, it's always a good thing to know how and why a project started.

I started exadrums because all the drum modules I knew looked the same: a small lcd display, buttons everywhere, and they were big devices. And they also were quite expensive!
Another reason is that commercial drum modules are basically a black box, you don't know what's inside, and I wanted to know.
More than that, I wanted to know if I could make a drum module that would be a bit more "modern".

So because exadrums is a complete drum module (i.e. its output is sound, not midi), I thought about a tablet-like device, which is why I went for a 7" touchscreen. In 2015, I was a bit ahead of Pearl, because the Mimic Pro was released more than a year later. But the Mimic is huge, and expensive, so that kept me very motivated.

After spending a lot of time on the software and the trigger board (3 years!), I was finally able to replace the TD-4 module by the exadrums module and play drums on my TD-4KP kit, with some limitations. You can read more about that [here](https://news.exadrums.com/article/how-to-make-a-drum-module), if you're interested.

The next step is now to make a bigger 16-input trigger board, to use exadrums with my TD-17KV. The challenge will be to keep exactly the same form factor, and there's still a lot of work to be done on the software.
So, hopefully, next year I'll replace my TD-17 module, by the module I've made!

**Reply by corrados** [reply]
> Another reason is that commercial drum modules are basically a black box, you don't know what's inside, and I wanted to know.

Ah, yes. Good point. I forgot about this. If you build your own module, you have control over everything and can fine tune the algorithms without being limited to the parameter set provided by the commercial module manufacturer.

> I was finally able to replace the TD-4 module by the exadrums module and play drums on my TD-4KP kit,

What is your impression, does exadrums performs similar to the TD-4 module or even better? Right now, Edrumulus performs worse than the Roland TD-27. But the TD-27 is a very advanced edrum module so, comparing a DIY module to this is a bit unfair :-).

**Reply by SpintroniK** [reply]
> > I was finally able to replace the TD-4 module by the exadrums module and play drums on my TD-4KP kit,
> 
> What is your impression, does exadrums performs similar to the TD-4 module or even better? Right now, Edrumulus performs worse than the Roland TD-27. But the TD-27 is a very advanced edrum module so, comparing a DIY module to this is a bit unfair :-).

It depends on what feature we want to compare. I would say exadrums is nearly as good as the TD-4KP in terms of triggering, which is not that bad to be fair (the TD-4KP is not bad at all).
I feel like the triggering is as good, or even better than the TD-6V, but that's quite subjective.

In terms of latency, a Raspberry Pi will never beat a Roland module (even a Pi4 with the best audio DAC), but it can get close.
I would say that the latency is not noticeable even with a scan time of 4ms.
As for the sound quality, it's better than the TD-4KP, because it's studio recorded16-bit PCM @ 48kHz, and both the TD-6V and TD4-KP sound a bit like fake sounds. The TD-17KV though, sounds almost as good, but triggering and latency are better than exadrums.

Now, if we talk about the user interface, I prefer the touchscreen, but not everyone will agree...

**Reply by corrados** [reply]
> I would say exadrums is nearly as good as the TD-4KP in terms of triggering, which is not that bad to be fair

Sounds good. As far as I understand your code, you do not implement a retrigger cancellation as the TD--4 does. Looking at [this file](https://github.com/SpintroniK/libeXaDrums/blob/master/DrumKit/Triggers/Triggers/DiscreteTrigger.cpp) you seem to remove the DC offset and then search for the maximum in the scan time. My impression is that without a retrigger cancellation, the fast press rolls cannot be triggered correctly. But maybe with your design it still works fine.

> In terms of latency, a Raspberry Pi will never beat a Roland module

Yes, the latency of the Roland modules is very impressive. I think they have the best latency on the market. Very hard to beat :-).

> As for the sound quality, it's better than the TD-4KP, because it's studio recorded16-bit PCM @ 48kHz

Do you use multiple samples for each instrument (like the snare)? With Edrumulus I am using the Drumgizmo software and with their [Crocellkit](https://drumgizmo.org/wiki/doku.php?id=kits:crocellkit), they use, e.g., almost 100 samples for the snare.

**Reply by SpintroniK** [reply]
> Sounds good. As far as I understand your code, you do not implement a retrigger cancellation as the TD--4 does. Looking at [this file](https://github.com/SpintroniK/libeXaDrums/blob/master/DrumKit/Triggers/Triggers/DiscreteTrigger.cpp) you seem to remove the DC offset and then search for the maximum in the scan time. My impression is that without a retrigger cancellation, the fast press rolls cannot be triggered correctly. But maybe with your design it still works fine.
> 
Well, there are indeed two features that are missing and could improve the triggering: retrig cancellation and crosstalk cancellation.
I feel like the retrig isn't really need, at least I haven't really noticed any problems even with fast rolls. Perhaps the mask time does the job just fine. I get a bit of retrig on the hi-hat, but it doesn't work well even with the TD-4 module, so I think it's reaching its end of life.
The crosstalk issue is noticeable when I use the bass drum, sometimes the floor tom triggers. It's possible to eliminate that completely, but it takes some time. So it would be nice to add crosstalk cancellation, for sure! Retrig cancellation would probably also improve things, especially for the hi-hat.

> 
> Do you use multiple samples for each instrument (like the snare)? With Edrumulus I am using the Drumgizmo software and with their [Crocellkit](https://drumgizmo.org/wiki/doku.php?id=kits:crocellkit), they use, e.g., almost 100 samples for the snare.

So, because it's using a Raspberry Pi, there's basically no limitation on the number of samples an instrument can play.
The way it's done right now is that, you can have as many samples per instrument as you want, but you have to write the instrument code yourself, re-compile, etc. For instance, the current hi-hat implementation uses 10 sounds: https://github.com/SpintroniK/libeXaDrums/blob/32a4045adf37e9f3b6a53e80543c9fbc053ea3f7/DrumKit/Instruments/Cymbals/TestHiHat.cpp#L68 but you could have a hundred, by changing the "10" into a "100" in that for loop.

What you're pointing at are the next two big features I'm working on to go from a very simple, proof of concept, TD-4-like module to a more advanced TD-17-like module.
The first feature is: support more trigger types (chokes, hi-hat, etc.), and improve the triggers by the mean time (which means that some work has to be done on the trigger input board as well).
The second one is: make the instruments customizable without having to code a line of C++, nor re-compile the code.
That'll be next year's big tasks.

**Reply by corrados** [reply]
> I feel like the retrig isn't really need, at least I haven't really noticed any problems even with fast rolls. Perhaps the mask time does the job just fine.

Yes, the mask time can help. But you may get missed triggers if the time is too large. But maybe with a fast roll you have so many hits that it does not matter if some of them are missing.

> The crosstalk issue is noticeable when I use the bass drum, sometimes the floor tom triggers.

Same with Edrumulus. I just recently added an experimental crosstalk cancellation since I got false triggers of the crash caused by the kick pad. It's definitely worth the effort to implement this.

> support more trigger types (chokes, hi-hat, etc.), and improve the triggers by the mean time [...] make the instruments customizable without having to code a line of C++, nor re-compile the code. That'll be next year's big tasks.

I hope it works out the way you intend it to be.

I think Edrumulus and exadrums basically do the same but with a very different approach. Your approach is to use an external ADC board and do everything else on the Raspi whereas Edrumulus make use of all the ADCs available on microcontrollers like the ESP32 or Teensy. So, Edrumulus is a pure trigger module and exadrum is a complete sound module.

**Reply by SpintroniK** [reply]
> Yes, the mask time can help. But you may get missed triggers if the time is too large. But maybe with a fast roll you have so many hits that it does not matter if some of them are missing.
> 
> 

That doesn't seem to be the case, but yeah, it'b be best to have that feature. Plus, it's not too hard to implement.
> 

> 
> Same with Edrumulus. I just recently added an experimental crosstalk cancellation since I got false triggers of the crash caused by the kick pad. It's definitely worth the effort to implement this.
> 
> 

I'll have a look at your code. I already have a good way to implement that.
> 
> 

> 
> I hope it works out the way you intend it to be.
> 
> 

Thanks, I think it will. At least my tests are very encouraging.

> 
> I think Edrumulus and exadrums basically do the same but with a very different approach. Your approach is to use an external ADC board and do everything else on the Raspi whereas Edrumulus make use of all the ADCs available on microcontrollers like the ESP32 or Teensy. So, Edrumulus is a pure trigger module and exadrum is a complete sound module.

Yes, that's exactly what it is. It feels like the two projects have a lot in common!

**Comment by ignotus666** [comment]
I lack the knowledge or skills to design my own drum module, but can use a soldering iron and hand tools.

My main motivation is being able to use something I made myself. I thoroughly enjoy the whole process of building something and seeing it work (sometimes after a _lot_ of trial and error!). Saving cash is a secondary concern, and if you factored in all the hours I've spent tinkering over the years, I could probably have bought several top-of-the-range kits. But that's no fun. I also like to learn new things and building my kit was the gateway to learning to read schematics, program MCUs, etc. Over time I've even taught myself to design and build my own simple MIDI controllers and it all started with... a piezo.

**Reply by corrados** [reply]
After tweaking the configuration/parameters of your set for a long time, you have gathered a lot of knowledge about pads and triggering. Once you have build a prototype hardware running the Edrumulus software, I am curious how you like the performance. Right now I am the only one who uses it. It would be interesting to get feedback from you as you have so much experience with DIY edrums.

**Reply by ignotus666** [reply]
If it's ok I'll wait a until the project matures a bit and see what board you settle on (teensy or ESP32). I'll be glad to give you feedback. When you're confident you've got a working prototype you could also open a thread in the DIY section of vdrums.com - it's not as active as it used to be but there are still some DIY diehards hanging out there who might be interested.

**Reply by corrados** [reply]
> If it's ok I'll wait a until the project matures a bit and see what board you settle on

Of course :-). Still a lot todo on my list. It will take me some time to fix all this.

**Comment by thijstriemstra** [comment]
my 2 cents:

Our drummer uses an spd-30 (roland octopad v2) and, as a producer, I'd like to fiddle around with e-drums as well, but not spend > 500 euro to get a drum module to setup a comparable kit (hihat, snare, maybe kick, etc). So i bought a roland cy-5 and some cheap, 2nd hand pd-5 pads and was planning to get a [eDrumin](https://www.audiofront.net/eDrumIn.php) module to connect it to Ableton (with Superior Drummer 3 VST plugin) but I work with esp32 all day, so a open source solution like this makes much more sense and fun.

It'll probably end costing more time and energy than buying a module, but designing everything, inc the case, and making it work, is definitely worth it to me.

**Comment by corrados** [comment]
I recently bought a used Roland HD-1 drum kit for 50 €:
![grafik](https://user-images.githubusercontent.com/46655886/143074207-5c26d907-8d7a-4b17-baad-65deeb85201e.png)
It has CY-5 and PDX-8 pads. Unfortunately, these are not the original stereo pads but modified versions for the HD-1 drum kit which are mono, i.e. no rim shot and no choke function. But I hope that the main sensor is the same as in the original pads so that I can soon add support for these types of pads in Edrumulus.

**Reply by thijstriemstra** [reply]
great to see we might get cy-5 tested and supported!

**Comment by jmccart2** [comment]
I'm not sure how I stumbled on this project, but it looks great. I play the drums and tinker around with electronics, so I'd really like to try this out. I don't have any teensy boards at the moment, but I have a few ESP32 (S3, 32E I think). 
As for drums, I don't currently own an edrum kit, but I have a PDX-100 and a KT-10 hooked up to an SPD-SX. I really want the ability to add more pads, a hi-hat controller and positional sensing - all of which the SPD-SX can't do.

Although I lack a lot of coding skills, I have a lot of electronic components and tools at my disposal. 
I'm willing to help in any way I can!

**Reply by corrados** [reply]
> I don't have any teensy boards at the moment, but I have a few ESP32 (S3, 32E I think).

Right now I am trying to get my prototype kit based on a Roland HD-1 set working correctly. I am using an ESP32 for that prototype and it looks promising. The Teensy has its advantages over the ESP32 but the ESP32 still seems to be a good choice for Edrumulus. Make sure to use a dual-core ESP32 processor.

> I have a PDX-100 [...]  really want [...] positional sensing

I have a PD-120 which should be very similar to that pad. So, I hope that the positional sensing will work fine for that pad. thijstriemstra has a DrumTec Diabolo tom pad with a center mounted piezo and in theory the positional sensing should work just fine. But he sent me some recordings of that pad and it looks like the Edrumulus positional sensing algorithm does not work ok for that pad. I have no idea why it does perform so badly. I have a PD-120 and a PD-80R mesh pad and for both of them the positional sensing works just fine.

> I'm willing to help in any way I can!

Great, welcome to the DIY e-drum community :-).
